cirq-core/cirq/sim/simulation_engine.py

```diff
@@ -0,0 +1,238 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Batteries-included class for Cirq's built-in simulators."""
+
+import abc
+import collections
+from typing import (
+    Any,
+    Dict,
+    Iterator,
+    List,
+    Tuple,
+    TYPE_CHECKING,
+    cast,
+    Generic,
+    Type,
+)
+
+import numpy as np
+
+from cirq import circuits, ops, protocols, study, value, devices
+from cirq.sim.simulator import (
+    TStepResult,
+    TSimulationTrialResult,
+    TSimulatorState,
+    TActOnArgs,
+    SimulatesIntermediateState,
+    SimulatesSamples,
+    check_all_resolved,
+    split_into_matching_protocol_then_general,
+)
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class SimulationEngine(
```

Comment:
This could generate confusion with `cirq_google.Engine`. Maybe `SimulatorBase` instead?

Reply:
Done